### ISSUE TITLE
make slate a controlled input

### ISSUE THREAD

**Issue by macrozone** [ianstormtaylor/slate#5281]
**Problem**

Slate was recently changed to be an uncontrolled component. this leads to massive problems for integrators of slate as uncontrolled components are not predictable because they have an inner state that is not fully controllable from outside. 

Uncontrolled components therefore lead to bugs and inconsistencies

**Solution**

Slate should be a controlled component, means that it will always show consistently what is passed as a value

**Alternatives**

a workaround is shown here https://github.com/ianstormtaylor/slate/issues/4992 but that also does not always work. Since it relies on a `useEffect` the execution of the effect might be delayed and things get out of sync.

**Context**
I am maintaining https://github.com/react-page/react-page and we use slate for rich-text editing. Since the mentioned update we have a hard time to fix all inconsistencies regarding slate

**Comment by aboveyunhai** [comment]
Uncontrolled component itself its fine, the problem is that the person with knowledge who made those breaking changes, failed to include the `uncontrolled` doc for those changes, and neglect the `controlled behavior` in React. It starts to turn out to be a mess now.

One thing I am pretty sure is that if you are using `useEffect` to update the state without waiting something externally from outside of the app (like waiting for an api call and then update the internal data). You might be doing something wrong most of the times.

I believe the "intended" way based on the PR is to directly call those `Transforms.XXX, Editor.apply, etc.` to do the `state(editor value)` manipulation. You have to simplify, wrap those inner functions and expose to the users if you want to let user control the value since `slate-react` currently failed to provide some shortcut helper functions for this part.

**Comment by erano067** [comment]
Any update?

**Comment by 12joan** [comment]
I think keeping Slate as an uncontrolled input is the right choice, for reasons I'll go into below. But to summarise, the state requirements of a text editor are far too complex for a purely controlled approach to be useful. At the bottom of this comment, I present a proof of concept to navigate this complexity and make Slate partially (and usefully) controlled in the rare cases in which this truly the most natural solution.

**What state does Slate use?**
The most obvious state Slate uses is `editor.children`, responsible for the content or "value" you see in your editor. If this were the only way in which Slate is stateful, then making Slate controlled would be simple; you would just need to expose a pair of props to pass in and set the value.

Unfortunately, text editors store more state than just their value. The most problematic examples for making Slate controlled are `editor.selection` and `editor.history`.

`editor.selection` stores the current selection in the form of a range. It must either be `null` (in which case no part of the editor is selected), or it must reference a pair of valid points in the editor. A point, as the docs state, is a specific text node and an offset within that text node. Slate will crash if `editor.selection` references a point that does not exist. When making changes using operations, the selection is updated automatically.

`editor.history` (if we're using the history plugin) stores two stacks of operations: an undo stack and a redo stack. An operation, for those who aren't familiar with the inner workings of Slate, describes an atomic, reversible change to the value of the editor. When the user makes any change to the content of the editor, such as by typing, these changes are applied (via the `editor.apply` method) using operations. Each operation is also copied onto the undo stack. When the user performs an undo operation, a batch of operations is popped off the undo stack, inverted to produce the opposite effect, applied to the editor, and then pushed onto the redo stack. For this process to work without error, all changes to the editor's value must happen through operations.

**Why do these present an obstacle to making Slate controlled?**
The most obvious way of making Slate controlled is to directly replace `editor.children` with a new value and re-render the Slate component. Unfortunately, this breaks the rules that selection and history depend on.

Firstly, selection. Suppose the user's cursor is inside a particular text node, and we replace `editor.children` with a new value in which that text node does not exist, or where that text node now has a length less than the offset of the user's selection in the old text node. In either case, `editor.selection` is now pointing to a non-existent point. Since Slate cannot resolve this point to a DOM selection, Slate must crash.

When modifying the editor content using operations, these issue do not occur. This is because Slate intelligently modifies the selection in response to operations. For example, suppose the user's cursor, denoted by the symbol '|', is inside a text node like this: "Hello w|orld". The offset of the selected point is 7 (the length of "Hello w"). Now suppose we use an operation to delete the word "Hello" and the following space from this text node. The resulting text "world" has length 5, so our old offset of 7 no longer exists. Slate corrects for this by subtracting from the offset the length of the text that was removed, 6. Thus, our new offset is 1: "w|orld". To the user, it appears as if the selection has not moved, as it is still between the 'w' and the 'o' in "world".

Note that the above correction is only possible because Slate knows the exact nature of the change that occured to the editor's value. This is why operations must be used when modifying the editor's value to ensure that the selection remains correct.

Secondly, history. Suppose we start with a text node "Hello world" again (the selection is not important), and the user adds an '!' to the end to make "Hello world!" This insertion is made by way of an operation that looks something like this: `{ type: 'insert_text', path: [0, 0], offset: 11, text: "!" }`. This operation is pushed onto the undo stack so that the user can undo it if they so choose. Now suppose we directly modify `editor.children` to replace the text node with "Hello, world!", note the addition of a comma.

If the user now performs an undo, the `insert_text` operation from earlier will be inverted to produce `{ type: 'remove_text', path: [0, 0], offset: 11, text: "!" }`, and this operation will be applied to the editor. The default implementation of `remove_text` doesn't care about the exact contents of its `text` property, only `text.length`. Thus, it does not matter that due to the addition of the comma earlier, the character at offset 11 is now 'd' instead of '!'; Slate will happily remove it as per the operation. The confused user will now be left with "Hello, worl!", note the lack of a 'd'.

**Okay, but I still want to make external changes to my Slate editor**
One solution is to erase the problematic `editor.selection` and `editor.history` when the value of `editor.children` is modified. For many use cases, this may be appropriate, but it comes with two unfortunate side effects: the user's selection will be lost, and so will their undo history.

What we really want is to use operations to transform the existing `editor.children` to our new desired value, but avoid having to think in terms of operations when it's more natural to treat Slate as a controlled component whose value we can just change.

**A proof of concept solution**
Luckily, @pubuzhixing8 found a way of inferring the required operations to transform one Slate value into another in their [slate-diff](https://github.com/pubuzhixing8/slate-diff) library. I came across this library when working on a rich-text diffing solution for Plate, and I can attest that for simple changes, the algorithm can be made to be very reliable with a few modifications.

Using `slate-diff`, we can take the existing `editor.children` and our desired `value` and compute a set of operations that will get us from one to the other while keeping selection and history intact. I implemented a proof of concept for this here: https://github.com/ianstormtaylor/slate/compare/6548197...12joan:slate:proof-of-concept/controlled-slate. There are a few edge cases and performance issues, but those could be ironed out with sufficient effort.

https://github.com/user-attachments/assets/cea650ca-cae4-4d87-8f75-557fbacd0fbc

**Comment by kettanaito** [comment]
Huge +1 on this. Slate makes an incorrect assumption that all editor state lives within the editor. Whereas on practice, you are likely to have external state, like a Document, that stores and updates the content of the thing you're editing. The lack of controlled pattern makes it unnecessarily hard to use proper React patterns with Slate since `initialValue` accepts no updates by design. 

I recommend adhering to established React practices:

- `defaultValue` for uncontrolled behaviors when you need them;
- `value` and `onChange` for controlled behaviors.

**Comment by Georgegriff** [comment]
This make integrating really challenging - especially with common use cases like streaming text in from else where.

Is there no good way of doing this? manually setting editor.children doesn't work very reliably.

I've looked at the diffing solutions and they look like they work but there's some issues, i would try to look at the plate solution but i don't love having to adopt plate to solve this problem

**Comment by aboveyunhai** [comment]
> This make integrating really challenging - especially with common use cases like streaming text in from else where.
> 
> Is there no good way of doing this? manually setting editor.children doesn't work very reliably.
> 
> I've looked at the diffing solutions and they look like they work but there's some issues, i would try to look at the plate solution but i don't love having to adopt plate to solve this problem

you need to manually do something like `editor.insertXXXX(content)` to granularly control the change. You can't think it like a typical React thing, this is something where React is not feasible, treat it like a `<canvas />`.
cuz if you treat it like a single complex React state, if you just want to update one item out of the entire document, now the whole document will re-render upon every single change (state change), all the cursor/stroke position, scroll position, focus status, etc. may lose. 
That is also a huge performance issue.

fwiw, several years ago, I don't understand why the text editor moved toward `uncontrolled`, until I truly work on text-editor on prod for a while. 
You will likely have the same question for every single React text-editor library on the market. They are all uncontrolled.

**Comment by 12joan** [comment]
> i would try to look at the plate solution but i don't love having to adopt plate to solve this problem

To clarify, Plate doesn't support this either. I came across the diffing technology that can theoretically mitigate this problem while working on Plate, which is why I mentioned it.

**Comment by Georgegriff** [comment]
Yeah i tried diffing out  from here https://github.com/pubuzhixing8/slate-diff/tree/main/src

it worked-ish, it struggled to diff complex nested children properly, so for now i've ignored it.

My use case:
An LLM streaming text into the editor

My solution for now is, i already have a customRef on my slate editor with functions like clear, select, i've added a replace function:

```ts
 replace: newValue => {
        const deserializedValue = options.deserialize
          ? options.deserialize(newValue)
          : (newValue as Descendant[]);

        Transforms.delete(editor, {
          at: {
            anchor: Editor.start(editor, []),
            focus: Editor.end(editor, []),
          },
        });
        // now insert the new value
        withoutNormalizing(editor, () => {
          Transforms.insertFragment(editor, deserializedValue);
        });
      },
```
i might need to wrap this in a withoutMerging too for history, potentially.

I still need to experiment a bit more, but the thinking is for this use case i'm fine to just clear all and lose users selection its an insert from external source, still need to test, hoping it will work. 

I initially was quite hopeful for the diffing algorithm stuff, i might dive deeper into exactly why it didn't quite work once i have my replace method working.
